src/utils/styleProps.ts

```diff
@@ -28,23 +29,32 @@ function getValueByIndex(prop, index = 0) {
 	return prop;
 }
 
+interface IStylePropConfig {
+	/** Name of the style prop */
+	styleProp: string;
+	/** Transforms the prop value into a valid CSS property value. */
+	getValue?: (val: string, props: {[key: string]: any}) => any;
+	/** The CSS properties that the value should be applied to */
+	properties?: string[];
+	/** Use this instead of stylePropConfig.properties and stylePropConfig.getValue
+	 * to define more complex style props that need to generate multiple CSS rules.
+	 * Will be called with any value, so you have to make sure to return a falsy
+	 * value if you don't want to add a rule.
+	 */
+	getRules?: (prop: string, theme?: any) => {[key: string]: any}
```

Comment:
One thing bugging me a bit is that these are all optional, but _actually_ they aren't.

Is there a way in TS to say: Either use both `getValue` and `properties` (but not `getRules`), or use `getRules` (but not `getValue` or `properties`). This would then correctly reflect how the configs can be structured.

For example, using `getRules` and `properties` together should be invalid.

Reply:
The problem is they are optional when they are used. if you look at `displayProps.ts`, you can see in `displayPropsDef` it is possible to have an undefined value for getValue when it calls `console.warn()` . TS catches this stuff so even if we think certain things always have to be there, it can catch when things can sometimes be undefined.

Reply:
> Is there a way in TS to say: Either use both `getValue` and `properties` (but not `getRules`), or use `getRules` (but not `getValue` or `properties`). This would then correctly reflect how the configs can be structured.

For this we'd probably want to create multiple types of interfaces and use that where appropriate.

Reply:
> you can see in displayPropsDef it is possible to have an undefined value for getValue when it calls console.warn(). TS catches this stuff so even if we think certain things always have to be there, it can catch when things can sometimes be undefined.

Hm, that check only runtime-validates the value of the prop (e.g. to make sure people don't pass typos `display="bolck"` to a component). Once we have the proper interface set up we can probably get rid of it.

Apart from that, `getValue` (hopefully) won't ever get called when a prop is `undefined`.